plugins/amdgpu/amdgpu_plugin_util.c

```diff
@@ -0,0 +1,222 @@
+struct tp_system src_topology;
+struct tp_system dest_topology;
+
+/* Helper structures to encode device maps during Checkpoint and Restore operations */
+struct device_maps checkpoint_maps;
+struct device_maps restore_maps;
+
+int fd_next;
+
+bool kfd_fw_version_check;
+bool kfd_sdma_fw_version_check;
+bool kfd_caches_count_check;
+bool kfd_num_gws_check;
+bool kfd_vram_size_check;
+bool kfd_numa_check;
+bool kfd_capability_check;
+
+bool checkpoint_is_complete()
+{
+	return (dev_file_cnt == 0);
+}
+
+void decrement_checkpoint_count()
+{
+	dev_file_cnt--;
+	pr_info("\n");
+	pr_info("Sairam: %s(),  Number of Checkpoints LEFT: %d\n", __func__, dev_file_cnt);
+}
+
+void init_gpu_count(struct tp_system *topo)
+{
+	if (dev_file_cnt != 0)
+		return;
+
+	// We add ONE to include checkpointing of KFD device */
+	dev_file_cnt = 1 + topology_gpu_count(topo); 
+	pr_info("\n");
+	pr_info("Sairam: %s(),  Number of Checkpoints: %d\n", __func__, dev_file_cnt);
+}
+
+int read_fp(FILE *fp, void *buf, const size_t buf_len)
+{
+	size_t len_read;
+
+	len_read = fread(buf, 1, buf_len, fp);
+	if (len_read != buf_len) {
+		pr_perror("Unable to read file (read:%ld buf_len:%ld)", len_read, buf_len);
+		return -EIO;
+	}
+	return 0;
+}
+
+int write_fp(FILE *fp, const void *buf, const size_t buf_len)
+{
+	size_t len_write;
+
+	len_write = fwrite(buf, 1, buf_len, fp);
+	if (len_write != buf_len) {
+		pr_perror("Unable to write file (wrote:%ld buf_len:%ld)", len_write, buf_len);
+		return -EIO;
+	}
+	return 0;
+}
+
+/**
+ * @brief Open an image file
+ *
+ * We store the size of the actual contents in the first 8-bytes of
+ * the file. This allows us to determine the file size when using
+ * criu_image_streamer when fseek and fstat are not available. The
+ * FILE * returned is already at the location of the first actual
+ * contents.
+ *
+ * @param path The file path
+ * @param write False for read, true for write
+ * @param size Size of actual contents
+ * @return FILE *if successful, NULL if failed
+ */
+FILE *open_img_file(char *path, bool write, size_t *size)
+{
+	FILE *fp = NULL;
+	int fd, ret;
+
+	if (opts.stream)
+		fd = img_streamer_open(path, write ? O_DUMP : O_RSTR);
+	else
+		fd = openat(criu_get_image_dir(), path, write ? (O_WRONLY | O_CREAT) : O_RDONLY, 0600);
+
+	if (fd < 0) {
+		pr_perror("%s: Failed to open for %s", path, write ? "write" : "read");
+		return NULL;
+	}
+
+	fp = fdopen(fd, write ? "w" : "r");
+	if (!fp) {
+		pr_perror("%s: Failed get pointer for %s", path, write ? "write" : "read");
+		return NULL;
+	}
+
+	if (write)
+		ret = write_fp(fp, size, sizeof(*size));
+	else
+		ret = read_fp(fp, size, sizeof(*size));
+
+	if (ret) {
+		pr_perror("%s:Failed to access file size", path);
+		fclose(fp);
+		return NULL;
+	}
+
+	pr_debug("%s:Opened file for %s with size:%ld\n", path, write ? "write" : "read", *size);
+	return fp;
+}
+
+int read_file(const char *file_path, void *buf, const size_t buf_len)
```

Comment:
Would it be possible to reuse `read_img_buf()` from `image.h`?

Reply:
Will consider doing so in next set of patches. Want to avoid another round of testing.